drivers/network/dd/b57xx/ndis.c

```diff
@@ -0,0 +1,557 @@
+                     &Adapter->StandardProducer.RingControlBlock.HostRingAddress);
+    
+    FreeSharedMemory(Adapter,
+                     sizeof(B57XX_RECEIVE_BUFFER_DESCRIPTOR) * Adapter->ReturnConsumer[0].Count,
+                     (PVOID)Adapter->ReturnConsumer[0].pRing,
+                     &Adapter->ReturnConsumer[0].RingControlBlock.HostRingAddress);
+    
+    FreeSharedMemory(Adapter,
+                     sizeof(B57XX_SEND_BUFFER_DESCRIPTOR) * Adapter->SendProducer[0].Count,
+                     (PVOID)Adapter->SendProducer[0].pRing,
+                     &Adapter->SendProducer[0].RingControlBlock.HostRingAddress);
+    
+    if (Adapter->SendProducer[0].pPacketList != NULL)
+    {
+        NdisFreeMemory(Adapter->SendProducer[0].pPacketList,
+                       sizeof(PNDIS_PACKET) * Adapter->SendProducer[0].Count,
+                       0);
+    }
+    
+    if (Adapter->Status.pBlock != NULL)
+    {
+        NdisMFreeSharedMemory(Adapter->MiniportAdapterHandle,
+                              sizeof(B57XX_STATUS_BLOCK),
+                              FALSE,
+                              (PVOID)Adapter->Status.pBlock,
+                              Adapter->Status.HostAddress);
+    }
+    
+    if (Adapter->Statistics.pBlock != NULL)
+    {
+        NdisMFreeSharedMemory(Adapter->MiniportAdapterHandle,
+                              sizeof(B57XX_STATISTICS_BLOCK),
+                              FALSE,
+                              (PVOID)Adapter->Statistics.pBlock,
+                              Adapter->Statistics.HostAddress);
+    }
+    
+    /* Free MMIO */
+    if (Adapter->IoBase)
+    {
+        NdisMUnmapIoSpace(Adapter->MiniportAdapterHandle, Adapter->IoBase, Adapter->IoLength);
+    }
+
+    /* Destroy the adapter context */
+    NdisFreeMemory(Adapter, sizeof(B57XX_ADAPTER), 0);
+}
+
+NDIS_STATUS
+NTAPI
+MiniportSend(_In_ NDIS_HANDLE MiniportAdapterContext,
+             _In_ PNDIS_PACKET Packet,
+             _In_ UINT Flags)
+{
+    NDIS_STATUS Status;
+    PB57XX_ADAPTER Adapter = (PB57XX_ADAPTER)MiniportAdapterContext;
+    PSCATTER_GATHER_LIST SGList = NDIS_PER_PACKET_INFO_FROM_PACKET(Packet,
+                                                                   ScatterGatherListPacketInfo);
+
+    
+    NdisAcquireSpinLock(&Adapter->SendLock);
+    
+    ASSERT(SGList != NULL);
+    
+    for (UINT i = 0; i < SGList->NumberOfElements; i++)
+    {
+        ASSERT(SGList->Elements[i].Length <= Adapter->MaxFrameSize);
+        
+        Adapter->SendProducer[0].pPacketList[Adapter->SendProducer[0].ProducerIndex] = Packet;
+
+        Status = NICTransmitPacket(Adapter,
+                                   SGList->Elements[i].Address,
+                                   SGList->Elements[i].Length,
+                                   i == SGList->NumberOfElements - 1 ? B57XX_SBD_PACKET_END : 0);
+        if (Status != NDIS_STATUS_SUCCESS)
+        {
+            Adapter->Statistics.TransmitErrors++;
+        }
```

Comment:
```suggestion
```
Statistics does not depend on resources availability. You need to check a descriptor flag status or dump hardware counters.